index.js

```diff
@@ -1,22 +1,56 @@
-module.exports = Hook
-
 var register = require('./lib/register')
 var addHook = require('./lib/add')
 var removeHook = require('./lib/remove')
 
-function Hook () {
+// https://stackoverflow.com/a/21792913
+var bind = Function.bind
+var bindable = bind.bind(bind)
+
+function bindApi (hook, state, name) {
+  const removeHookRef = bindable(removeHook, null).apply(null, name ? [state, null, name] : [state, null])
+  hook.api = { remove: removeHookRef }
+  hook.remove = removeHookRef
+
+  ;['before', 'error', 'after', 'wrap'].forEach(function (kind) {
+    const args = name ? [state, kind, name] : [state, kind]
+    hook[kind] = hook.api[kind] = bindable(addHook, null).apply(null, args)
+    hook.remove[kind] = hook.api.remove[kind] = bindable(removeHook, null).apply(null, args)
+  })
+}
+
+function HookSingular () {
+  var singularHookName = 'h'
+  var singularHookState = {
+    registry: {}
+  }
+  var singularHook = register.bind(null, singularHookState, singularHookName)
+  bindApi(singularHook, singularHookState, singularHookName)
+  return singularHook
+}
+
+function HookCollection () {
   var state = {
     registry: {}
   }
 
   var hook = register.bind(null, state)
-  hook.api = { remove: removeHook.bind(null, state, null) }
-  hook.remove = removeHook.bind(null, state, null)
-
-  ;['before', 'error', 'after', 'wrap'].forEach(function (kind) {
-    hook[kind] = hook.api[kind] = addHook.bind(null, state, kind)
-    hook.remove[kind] = hook.api.remove[kind] = removeHook.bind(null, state, kind)
-  })
+  bindApi(hook, state)
 
   return hook
 }
+
+var collectionHookDeprecationMessageDisplayed = false
+function Hook () {
+  if (!collectionHookDeprecationMessageDisplayed) {
+    console.warn('[before-after-hook]: "Hook()" repurposing warning. Read more: https://github.com/gr2m/before-after-hook#upgrading-to-1.4 ')
+    collectionHookDeprecationMessageDisplayed = true
+  }
+  return HookCollection()
+}
+
+Hook.Singular = HookSingular.bind(null)
+Hook.Collection = HookCollection.bind(null)
```

Comment:
Why does it need binding to `null`?

Reply:
I guess it doesn't, unit tests still pass 👍